### DISCUSSION TITLE
Drayton Wiser TRVs

### DISCUSSION THREAD

**Discussion by jamesonuk** [Koenkk/zigbee2mqtt#9667] [Devices]
I have a few [Drayton Wiser TRVs](https://www.zigbee2mqtt.io/devices/WV704R0A0902.html#schneider%2520electric-wv704r0a0902) which are linked to Zigbee2MQTT via a Conbee II stick.

First question is a simple one.   I have seen in the debug logs something about the TRV calling for OTA.   The devices don't appear on the OTA page so I am assuming that updates are not supported.   Is there anything I can do to help get these devices to update via Zigbee2MQTT?

Secondly the docs state:
> * system_mode: Mode of this device. To control publish a message to topic zigbee2mqtt/FRIENDLY_NAME/set with payload {"system_mode": VALUE} where VALUE is one of: off, auto, heat. To read send a message to zigbee2mqtt/FRIENDLY_NAME/get with payload {"system_mode": ""}.
> * running_state: The current running state. Possible values are: idle, heat. To read send a message to zigbee2mqtt/FRIENDLY_NAME/get with payload {"running_state": ""}.

But looking at the state of the device I only ever see null for `running_state` and no `system_mode`
![image](https://user-images.githubusercontent.com/1040621/141751336-f58f4e2c-16d1-4e07-a182-3095723366e5.png)
This seems to cause some warnings in HA as it looks like `system_mode` is being sent through as null.

**Comment by sjorge** [comment]
If you manage to get hold of the firmware files, you can use the scripts here https://github.com/Koenkk/zigbee-OTA to add them to the repo. Then devices should be able to get the update. I think you might also need to update the device config to tell it it can do OTA updates. I think the default is to just not check.

Not sure for the running state, is there any error logged when sedding the get for the running_state over mqtt?

**Reply by jamesonuk** [reply]
> If you manage to get hold of the firmware files, you can use the scripts here https://github.com/Koenkk/zigbee-OTA to add them to the repo. Then devices should be able to get the update. I think you might also need to update the device config to tell it it can do OTA updates. I think the default is to just not check.

Just had a look and see now what needs doing (just include the ota file and add `ota: ota.zigbeeOTA,` to the device.  It currently errors with no details for image type 513 but that is obviously the fact I don't have the firmware available to update that side of things.
> Not sure for the running state, is there any error logged when sedding the get for the running_state over mqtt?
The screenshot above is from Zigbee2MQTT frontend before it is in MQTT.
If I send a get message from HA that is included in the logs below but I can't see anything else of use

```
Info MQTT publish: topic 'zigbee2mqtt/Bedroom TRV', payload '{"ADC":"1514,1517,2,175,-32768,1900,0,500,100,175,174,178","ALG":"8eb,1,190,175,15,0,500,730,14822,100,100","MOT":"CloseValve","battery":100,"keypad_lockout":"unlock","last_seen":"2021-11-15T14:20:09.725Z","linkquality":255,"local_temperature":17.5,"occupied_heating_setpoint":19,"pi_heating_demand":39,"running_state":null,"update_available":null,"voltage":3100}'
Info MQTT publish: topic 'zigbee2mqtt/Bedroom TRV', payload '{"ADC":"1513,1517,3,175,-32768,1900,0,500,100,175,174,177","ALG":"8eb,1,190,175,15,0,500,730,14822,100,100","MOT":"CloseValve","battery":100,"keypad_lockout":"unlock","last_seen":"2021-11-15T14:20:09.725Z","linkquality":255,"local_temperature":17.5,"occupied_heating_setpoint":19,"pi_heating_demand":39,"running_state":null,"update_available":null,"voltage":3100}'
Debug Received Zigbee message from 'Bedroom TRV', type 'attributeReport', cluster 'wiserDeviceInfo', data '{"deviceInfo":"ALG,8ec,1,190,175,15,0,500,730,14822,100,100"}' from endpoint 1 with groupID null
Info MQTT publish: topic 'zigbee2mqtt/Bedroom TRV', payload '{"ADC":"1513,1517,3,175,-32768,1900,0,500,100,175,174,177","ALG":"8ec,1,190,175,15,0,500,730,14822,100,100","MOT":"CloseValve","battery":100,"keypad_lockout":"unlock","last_seen":"2021-11-15T14:20:09.752Z","linkquality":255,"local_temperature":17.5,"occupied_heating_setpoint":19,"pi_heating_demand":100,"running_state":null,"update_available":null,"voltage":3100}'
Info MQTT publish: topic 'zigbee2mqtt/Bedroom TRV', payload '{"ADC":"1513,1517,3,175,-32768,1900,0,500,100,175,174,177","ALG":"8ec,1,190,175,15,0,500,730,14822,100,100","MOT":"CloseValve","battery":100,"keypad_lockout":"unlock","last_seen":"2021-11-15T14:20:09.752Z","linkquality":255,"local_temperature":17.5,"occupied_heating_setpoint":19,"pi_heating_demand":39,"running_state":null,"update_available":null,"voltage":3100}'
Debug Received Zigbee message from 'Bedroom TRV', type 'attributeReport', cluster 'hvacThermostat', data '{"57362":0,"57363":4,"57364":0,"57365":120}' from endpoint 1 with groupID null
Debug Received Zigbee message from 'Bedroom TRV', type 'attributeReport', cluster 'haDiagnostic', data '{"lastMessageLqi":100,"lastMessageRssi":-75}' from endpoint 1 with groupID null
Debug Received Zigbee message from 'Bedroom TRV', type 'attributeReport', cluster 'haDiagnostic', data '{"57345":10152}' from endpoint 1 with groupID null
Debug Received MQTT message on 'zigbee2mqtt/Bedroom TRV/get' with data '{running_state: ''}'
```

**Reply by sjorge** [reply]
No readResponse debug message after the /get ? That probably means it's not reading anything from the device.

**Reply by jamesonuk** [reply]
> No readResponse debug message after the /get ? That probably means it's not reading anything from the device.

Looking at the logs on the server not the frontend I see

```
debug 2021-11-15 14:18:53: Received MQTT message on 'zigbee2mqtt/Bedroom TRV/get' with data '{running_state: ''}'
error 2021-11-15 14:18:53: Invalid message 'null', skipping...
info  2021-11-15 14:19:07: MQTT publish: topic 'zigbee2mqtt/Bedroom TRV', payload '{"ADC":"1513,1516,3,175,-32768,1900,0,500,100,175,174,177","ALG":"8eb,1,190,175,15,0,
500,730,14822,100,100","MOT":"CloseValve","battery":100,"keypad_lockout":"unlock","last_seen":"2021-11-15T14:19:07.612Z","linkquality":255,"local_temperature":17.5,"occ
upied_heating_setpoint":19,"pi_heating_demand":39,"running_state":null,"update_available":null,"voltage":3100}'
```

**Reply by sjorge** [reply]
Interesting, I think the payload that got send is invalid `'{running_state: ''}'` I'd expect it to be `'{"running_state": ""}'`

**Reply by jamesonuk** [reply]
> Interesting, I think the payload that got send is invalid `'{running_state: ''}'` I'd expect it to be `'{"running_state": ""}'`

That was me running it manually :)

Should be
```
debug 2021-11-15 14:35:09: Bedroom TRV/Bedroom TRV/get' with data '{"running_state": ""}'
error 2021-11-15 14:35:09: No converter available for 'running_state' ()
```

**Reply by sjorge** [reply]
```
error 2021-11-15 14:35:09: No converter available for 'running_state' ()
```
OK so there is no (get) converter for running_state at least :(

Might be why it's always null if the device doesn't report it on it's own.

**Reply by sjorge** [reply]
Looking at the device there are some weird things there:
https://github.com/Koenkk/zigbee-herdsman-converters/blob/563d7b20e1a2ad3a509800b47be297f57b669c79/devices/schneider_electric.js#L11-L33

It has `fz.ignore_haDiagnostic` in fromZigbee but then it also explicitly binds that in the configure function...
```
const binds = ['genBasic', 'genPowerCfg', 'hvacThermostat', 'haDiagnostic'];
```

There should probably also not be a binding for `haDiagnostic` as we will be ignoring it, whats worse then later there is:
```
// bind of hvacUserInterfaceCfg fails with 'Table Full', does this have any effect?
```

But we do try and setup reporting for hvacUserInterfaceCfg and that one is needed for keypad_lockout. I'm guessing at the very least the haDiagnostic bind needs to go and be replaced with a hvacUserInterfaceCfg bind.

But that won't solve the running_state problem though, I wonder if this was just copy-pasted from other TRV and this device doesn;t suppor it.

I also spot 2 legacy fromZigbee converters, they probably also need to be reworked.
```
fz.legacy.wiser_thermostat, fz.legacy.wiser_itrv_battery
```

**Reply by jamesonuk** [reply]
> Looking at the device there are some weird things there:
This is where I start to get lost.   I don't really understand the Zigbee cluster / binding stuff and Javascript has a big block on my brain so I am feeling around in the dark here.

What can I do to actually debug this?   Is there a way to dump the raw Zigbee messages and something that might explain how to parse them...

**Comment by sjorge** [comment]
All the info seems to be in the ALG and ADC, but not sure how to parse.

~ sjorge

**Reply by sjorge** [reply]
You could look at https://github.com/dresden-elektronik/deconz-rest-plugin/issues/4148 for some hints I guess.

**Reply by sjorge** [reply]
Yep, looks like the cluster.zip has a bunch of useful info.

**Reply by jamesonuk** [reply]
> Yep, looks like the cluster.zip has a bunch of useful info.

I can see how that seems to map the raw Zigbee messages but I am struggling to figure out how they are parsed in Zigbee2MQTT.
I have been though the Herdsman Converters but I am guessing the actual parsing of messages is done in Herdsman?

**Reply by sjorge** [reply]
Parsing is done in zigbee-herdsman-converters, under devices/ directory there are the definition that link the zigbee model and other info with a set of converters that are called for messages received/send to the device.

These converters are in converters/fromZigbee.js and toZigbee.js depending on the direction of the comunication.

**Reply by jamesonuk** [reply]
> Parsing is done in zigbee-herdsman-converters, under devices/ directory there are the definition that link the zigbee model and other info with a set of converters that are called for messages received/send to the device.
> 
> These converters are in converters/fromZigbee.js and toZigbee.js depending on the direction of the comunication.

Which is where I got lost.   The system mode appears to come in as 0x001c but that doesn't appear as a value in the converters project at all so I was struggling to see how that inbound raw data could be parsed as
```
if (msg.data.hasOwnProperty('systemMode')) {
    result[postfixWithEndpointName('system_mode', msg, model)] = constants.thermostatSystemModes[msg.data['systemMode']];
}
```

**Reply by sjorge** [reply]
There are cluster/attribute definition that map the type, so there are pretty name in msg.data that contain the correct data (unintX, intX, string, ...)

The define for those is here: https://github.com/Koenkk/zigbee-herdsman/blob/master/src/zcl/definition/cluster.ts, e.g. systemMode specifically is here https://github.com/Koenkk/zigbee-herdsman/blob/8de234b76d9ed367916f45862f70693fcbb395cb/src/zcl/definition/cluster.ts#L1900

The type is enum8, so that contains a fixed set of value. Since this is a standard zigbee attribute, we can lookup what the enum values mean in the ZCL (there is a copy here https://github.com/Koenkk/zigbee-herdsman/blob/master/docs/Zigbee%20Cluster%20Library%20Specification%20v7.pdf)

Give this is a well known one, you can usually find it in lib/constants in zigbee-herdsman-conveters, e.g. https://github.com/Koenkk/zigbee-herdsman-converters/blob/7ee6dea2b9157f28259a6fab6d62c7405cd3eae7/lib/constants.js#L24-L34

**Reply by jamesonuk** [reply]
> The define for those is here: https://github.com/Koenkk/zigbee-herdsman/blob/master/src/zcl/definition/cluster.ts, e.g. systemMode specifically is here https://github.com/Koenkk/zigbee-herdsman/blob/8de234b76d9ed367916f45862f70693fcbb395cb/src/zcl/definition/cluster.ts#L1900

Ah I see it is decimal here which would explain why I couldn't find the hex value :)

**Reply by sjorge** [reply]
Yeah it's a mix of hex and decimal, most of the ZCL mentions hex so it's sometimes easy to miss something :(

**Comment by jamesonuk** [comment]
So is there a way I can dump the Zigbee messages so I can see what is being sent from the device?

**Reply by sjorge** [reply]
A 2nd zigbee stick with a sniffer firmware + wireshark can dump what is being send, do note you need to figure out the encryption key that is used if you want to get anything useful.

**Reply by jamesonuk** [reply]
But is some of this not already parsed by Zigbee Herdsman?
